is/utils/io.hpp

```diff
@@ -1,10 +1,64 @@
 #pragma once
 
 #include <string>
+#include <fstream>
 #include "google/protobuf/message.h"
 #include "status.hpp"
+#include "boost/optional.hpp"
+
+namespace google {
+namespace protobuf {
+namespace io {
+class ZeroCopyOutputStream;
+class CodedOutputStream;
+class CodedInputStream;
+}
+}
+}
 
 namespace is {
 wire::Status load(std::string const& filename, google::protobuf::Message* message);
 wire::Status save(std::string const& filename, google::protobuf::Message const& message);
+
+struct ProtobufWriter {
+  std::ofstream file;
+  google::protobuf::io::ZeroCopyOutputStream *raw_output;
+  google::protobuf::io::CodedOutputStream *coded_output;
+
+  ProtobufWriter(std::string const &filename);
+  ~ProtobufWriter();
+  void close();
+
+  template <typename T> bool insert(T const &message) {
+    coded_output->WriteVarint32(message.ByteSize());
+    return message.SerializeToCodedStream(coded_output);
+  }
+};
+
+struct ProtobufReader {
+  std::ifstream file;
+  google::protobuf::io::ZeroCopyInputStream *raw_input;
+  google::protobuf::io::CodedInputStream *coded_input;
+
+  ProtobufReader(std::string const &filename);
+  ~ProtobufReader();
+  void close();
+
+  template <typename T> boost::optional<T> next() {
```

Comment:
Same comments as in ProtobufWrite::insert. If you decide to return Status here change function signature to receive google::protobuf::Message* (and to the same for ProtobufWrite::insert)